modApiExt/alter.lua

```diff
@@ -389,128 +390,119 @@ function SpaceScript(loc, script)
-				self == Prime_Punchmech_AB
-			then
-				-- Add a dummy damage instance to fix Ramming Speed
-				-- achievement being incorrectly granted
-				fx:AddDamage(SpaceDamage(GetProjectileEnd(p1, p2)))
-			end
+		fx:AddScript(
+			"modApiExt_internal.fireSkillEndHooks("
+			.."modApiExt_internal.mission, Pawn,"
+			.."\""..self.__Id.."\","..p1:GetString()..","..p2:GetString()..")"
+		)
 
-			skillFx.effect = fx.effect
+		if
+			self == Prime_Punchmech    or
+			self == Prime_Punchmech_A  or
+			self == Prime_Punchmech_B  or
+			self == Prime_Punchmech_AB
+		then
+			-- Add a dummy damage instance to fix Ramming Speed
+			-- achievement being incorrectly granted
+			fx:AddDamage(SpaceDamage(GetProjectileEnd(p1, p2)))
 		end
 
-		if not skillFx.q_effect:empty() then
-			local fx = SkillEffect()
-			local effects = extract_table(skillFx.q_effect)
-
-			fx:AddScript(
-				"modApiExt_internal.fireQueuedSkillStartHooks("
-				.."modApiExt_internal.mission, Pawn,"
-				.."\""..self.__Id.."\","..p1:GetString()..","..p2:GetString()..")"
-			)
+		skillFx.effect = fx.effect
+	end
 
-			for _, e in pairs(effects) do
-				fx.effect:push_back(e)
-			end
+	if not skillFx.q_effect:empty() then
+		local fx = SkillEffect()
+		local effects = extract_table(skillFx.q_effect)
 
-			fx:AddScript(
-				"modApiExt_internal.fireQueuedSkillEndHooks("
-				.."modApiExt_internal.mission, Pawn,"
-				.."\""..self.__Id.."\","..p1:GetString()..","..p2:GetString()..")"
-			)
+		fx:AddScript(
+			"modApiExt_internal.fireQueuedSkillStartHooks("
+			.."modApiExt_internal.mission, Pawn,"
+			.."\""..self.__Id.."\","..p1:GetString()..","..p2:GetString()..")"
+		)
 
-			skillFx.q_effect = fx.effect
+		for _, e in pairs(effects) do
+			fx.effect:push_back(e)
 		end
-	end
 
-	return skillFx
-end
-
-function modApiExtHooks:overrideSkill(id, skill)
-	assert(skill.GetSkillEffect)
-	assert(_G[id] == skill) -- no fun allowed
+		fx:AddScript(
+			"modApiExt_internal.fireQueuedSkillEndHooks("
+			.."modApiExt_internal.mission, Pawn,"
+			.."\""..self.__Id.."\","..p1:GetString()..","..p2:GetString()..")"
+		)
 
-	if modApiExt_internal.oldSkills[id] then
-		error(id .. " is already overridden!")
+		skillFx.q_effect = fx.effect
 	end
 
-	modApiExt_internal.oldSkills[id] = skill.GetSkillEffect
-
-	-- Make it possible to identify skills with no ambiguity
-	skill.__Id = id
-	skill.GetSkillEffect = modApiExtGetSkillEffect
+	return skillFx
 end
 
 function modApiExtHooks:overrideAllSkills()
 	if not modApiExt_internal.oldSkills then
 		modApiExt_internal.oldSkills = {}
 
+		-- do this in two passes, so that for weapon upgrades we don't
+		-- accidentally set their original skill to our override, if we're
+		-- unlucky with iteration order.
 		for k, v in pairs(_G) do
 			if type(v) == "table" and v.GetSkillEffect then
-				self:overrideSkill(k, v)
+				v.__Id = k
+				originalSkillsEnv[k] = setmetatable(
+					{ GetSkillEffect = v.GetSkillEffect },
+					{ __index = skill }
```

Comment:
Where does `skill` come from, shouldn't that be `v`?

Reply:
You're right, good catch